lichess.js

```diff
@@ -0,0 +1,143 @@
+//------------------------------------------------------------------------------
+// Lichess helpers
+//------------------------------------------------------------------------------
+var http = require("./http.js");
+var moment = require("moment");
+var Q = require("q");
+
+var MILISECOND = 1;
+var SECONDS = 1000 * MILISECOND;
+var MINUTES = 60 * SECONDS;
+
+//------------------------------------------------------------------------------
+// A lichess API queue.
+//
+// Provides two queues, a background queue and main queue. Use the main queue
+// for any processing in response to a user message, as the main queue takes
+// priority.
+//
+// Use the background queue for any background updating, like player ratings.
+//
+// Also attempts to rate limit itself according to the lichess API guidelines
+// a max of one request every 2 seconds (this is actually a bit slower), and
+// it will wait a minute if it receives a 429.
+//------------------------------------------------------------------------------
+var makeRequest = (function() {
+    // The background queue will be processed when there is nothing else to process
+    var backgroundQueue = [];
+    // The mainQueue will be processed first.
+    var mainQueue = [];
+    var lastRequest = moment.utc();
+    function makeRequest(url, isJSON, isBackground) {
+        var deferred = Q.defer();
+        if (isBackground) {
+            backgroundQueue.push([url, isJSON, deferred]);
+        } else {
+            mainQueue.push([url, isJSON, deferred]);
+        }
+        return deferred.promise;
+    }
+    function processRequest() {
+        var request = null;
+        // The default delay between requests is 2 seconds
+        var requestDelay = 2 * SECONDS;
+        if (mainQueue.length > 0) {
+            request = mainQueue.shift();
+        } else if (backgroundQueue.length > 0) {
+            request = backgroundQueue.shift();
+        }
+        if (request) {
+            var url = request[0];
+            var isJSON = request[1];
+            var deferred = request[2];
+            var promise = null;
+            if (isJSON)  {
+                promise = http.fetchURLIntoJSON(url);
+            } else {
+                promise = http.fetchURL(url);
+            }
+            promise.then(function(result) {
+                if (result.response.statusCode == 429) {
```

Comment:
in case of a 429 in response to a user message, how much effort would it be to let the user know? is this information accessible from from that context? Just a potential enhancement. #58